packages/rs-dapi-client/src/dapi_client.rs

```diff
@@ -25,10 +33,40 @@ impl<TE: CanRetry> CanRetry for DapiClientError<TE> {
         match self {
             NoAvailableAddresses => false,
             Transport(transport_error) => transport_error.can_retry(),
+            #[cfg(feature = "mocks")]
+            MockExpectationNotFound(_) => false,
         }
     }
 }
 
+#[async_trait]
+/// DAPI client trait.
+pub trait Dapi {
+    /// Execute request using this DAPI client.
+    async fn execute<R>(
+        &mut self,
+        request: R,
+        settings: RequestSettings,
+    ) -> Result<R::Response, DapiClientError<<R::Client as TransportClient>::Error>>
+    where
+        R: TransportRequest;
+}
+
+#[async_trait]
+impl<'a, T: Dapi + Send> Dapi for MutexGuard<'a, T> {
```

Comment:
I think it's better to implement Dapi for a reference then, this should work for all smart pointers which MutexGuard belongs to

Reply:
Can you give me an example? Tried comment out this MutexGuard and to change `impl Dapi for DapiClient` to `impl Dapi for &DapiClient` (this tells me I cannot borrow-mut self.address_list),  and also `impl Dapi for &mut DapiClient` (this works but claims that MutexGuard does not implement Dapi).

Reply:
Do you need to explicitly pass MutexGuard somewhere? Because if not, MutexGuard<T> implements DerefMut and is applicable in places where D: Dapi works going through DerefMut (&mut T), so if we have `impl<D: Dapi> Dapi for &mut D` then anything that's capable of DerefMut/AsMut (both are ways to obtain &mut) could be converted to reference prior to function call

Reply:
Done in https://github.com/dashpay/platform/pull/1417/commits/e89c156e462d41c2e542076ede80c5b2e31a6a25, let me know if that's what you mean

Reply:
Exactly, thanks